proposals/0000-first-class-modules.rst

```diff
@@ -0,0 +1,612 @@
+
+That is, a qualified name is a simple name with a (possibly empty) list
+of qualifying names. These correspond to the dot-separated segments of
+the textual name.
+
+Scopes contain simple names only
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Currently, the ``inScope`` relation relates *qualified* names to
+entities, with the following signature:
+
+::
+
+   inScope :: Rel Module QualName Entity
+
+We propose that instead it relates *simple* names to entities, and we
+add a new relation ``refersTo`` which relates qualified names to
+entities in a scope. [#]_
+
+.. [#] By analogy to programming languages: the bindings in scope are like an environment,
+   simple names are like variables, qualified names are like expressions, and ``refersTo``
+   is ``evalExpr``.
+
+::
+
+   inScope :: Rel Module Name Entity
+   refersTo :: Rel Module QualName Entity
+
+``refersTo`` works as follows:
+
+- If the qualified name has no
+  qualifiers, then it refers to whatever that simple name maps to in the
+  scope.
+- If the qualified name has a qualifier, then we work out what
+  the qualifying name refers to. If it refers to a module, then we look at
+  the bindings exported by the module to find one matching the name.
+
+That is, we interpret ``A.b`` by first working out what ``A`` refers to, as a
+module, and then looking inside it to find ``b``. ``A`` may itself be
+qualified (since modules may export module bindings, just like any other
+bindings), hence the need for the recursive lookup.
+
+Exporting module names
+^^^^^^^^^^^^^^^^^^^^^^
+
+Module names *cannot* be exported, since there are no export specifiers that can
+refer to them. [#]_ The existing ``module M`` export specifier
+continues to work as it is defined in the Haskell report.
+
+.. [#] See `Module export specifiers`_ for the changes to allow this.
+
+Imports
+^^^^^^^
+
+Imports at the top level of a file module are treated specially. Their
+target is interpreted as a *dotted* name, which is resolved using the
+current rules for locating external Haskell modules. [#]_
+
+.. [#] For imports in other settings, see `Local modules`_.
+
+Importing a module brings all the exported bindings of the target module
+into scope, as described in [hsmods]_. It also brings
+the target module into scope bound to its declared simple name (which may be
+a dotted name if it is a file module).
+
+If a module is imported ``qualified`` then the target module is no
+brought into scope.
+
+If a module is imported ``as`` a simple name, then the target module is brought
+into scope with that simple name instead.
+
+If multiple modules are imported ``as`` the same name, then we retain the current
+behaviour, which is to “merge” modules. [#]_ The semantics is roughly that:
+
+.. [#] See `What should happen if multiple modules are defined
+       with the same name?`_ for discussion.
+
+::
+
+   import A as M
+   import B as M
+
+means the same as
+
+::
+
+   module M where
```

Comment:
Is this supposed to be `module M (module A, module B) where`? It's not clear to me (at this point in the proposal, maybe it's clarified later?) why importing `A` and `B` would re-export their contents from `M`.

Reply:
Ugh, yes, I was operating on an older model of mine where there were automatic re-exports. Yes, this should have something like `module (module A, module B)`, although I think I'll adopt your suggestion of `module A(..)`.